codec/src/codec.rs

```diff
@@ -40,17 +40,16 @@ pub trait Codec: Sized {
 }
 
 /// Trait for types that have a fixed-length encoding
-pub trait SizedCodec: Codec {
-    /// The encoded length of this value.
-    const LEN_CODEC: usize;
-
+pub trait SizedCodec<const N: usize>: Codec {
     /// Returns the encoded length of this value.
+    ///
+    /// Should not be overridden by implementations.
     fn len_encoded(&self) -> usize {
-        Self::LEN_CODEC
+        N
     }
 
     /// Encodes a value to fixed-size bytes.
-    fn encode_fixed<const N: usize>(&self) -> [u8; N] {
+    fn encode_fixed(&self) -> [u8; N] {
```

Comment:
This is safer now, you can't do stuff like `const buf: [u8; 3] = (1234 as u64).encode_fixed();` as it won't compile anymore due to that `3` not matching the length of `u64`

Wasn't able to figure out compiling tuples as fixed due to this change anymore but I think that's a fine tradeoff

Reply:
Agreed.

Reply:
I actually found an issue with this change so I will revert it.

This is called using a "generic parameter" rather than an "associated constant". And if `Array` trait has to implement this trait, it would cause `Array` to have to have a new generic parameter `N`. So `Array` becomes `Array<N>` which doesn't seem that bad, so far.

But then `Scheme` must then become: `Scheme<PRIVATE_LEN, PUBLIC_LEN, SIG_LEN>` with three different associated constants, and a `Config` that includes `Scheme` would also have to add three more associated constants, and so on. It starts to get really messy :confused: